src/helpers/proposed_network.rs

```diff
@@ -0,0 +1,198 @@
+
+    /// Returns a sink that accepts data to be sent to other helper parties.
+    fn sink(&self) -> Self::Sink;
+
+    /// Returns a stream to receive messages that have arrived from other helpers. Note that
+    /// some implementations may panic if this method is called more than once.
+    fn recv_stream(&self) -> Self::MessageStream;
+}
+
+pub struct CreateQueryData {
+    callback: oneshot::Sender<QueryId>,
+}
+
+impl CreateQueryData {
+    #[must_use]
+    pub fn new(callback: oneshot::Sender<QueryId>) -> Self {
+        CreateQueryData { callback }
+    }
+    pub fn respond(self, query_id: QueryId) -> Result<(), NetworkEventError> {
+        self.callback
+            .send(query_id)
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "CreateQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct PrepareQueryData {
+    pub query_id: QueryId,
+    callback: oneshot::Sender<()>,
+}
+
+impl PrepareQueryData {
+    #[must_use]
+    pub fn new(query_id: QueryId, callback: oneshot::Sender<()>) -> Self {
+        PrepareQueryData { query_id, callback }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "PrepareQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct StartMulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub field_type: String,
+    pub data_stream: ByteArrStream,
+    callback: oneshot::Sender<()>,
+}
+
+impl StartMulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        field_type: String,
+        data_stream: ByteArrStream,
+        callback: oneshot::Sender<()>,
+    ) -> Self {
+        StartMulData {
+            query_id,
+            endpoints_positions,
+            field_type,
+            data_stream,
+            callback,
+        }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "StartMul",
+                query_id,
+            })
+    }
+}
+
+pub struct MulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+}
+
+impl MulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+    ) -> Self {
+        Self {
+            query_id,
+            endpoints_positions,
+            endpoints_to_roles_and_stream,
+        }
+    }
+}
+
+pub struct RingEventData {
+    pub query_id: QueryId,
+    pub roles_to_endpoints: HashMap<Role, Uri>,
```

Comment:
can we wrap `Uri` into `HelperIdentity` struct?

Reply:
I think we should separate the concept of `Uri` and `HelperIdentity`. meaning, we have an opaque identity for helpers, and in addition to that, a `Uri` that the HTTP impl uses to communicate. then, we could have something like

```
HelperIdentity::from(uri: Uri)
HelperIdentity::from(str: &str)
etc...
```

internally, we could just `to_string` the `Uri` or something.

Reply:
i took a shot at implementing this

Reply:
hiding internals of `HelperIdentity` makes it an opaque ID as no component can look inside of it. But I think this is the right way to implement it